File: ios/TriPay/Domain/DomainValidator.swift
import Foundation

enum ValidationError: Error {
    case invalidAmount
    case noteTooLong
    case futureDate
}

struct DomainValidator {
    static func validate(transaction: Transaction) throws {
        if transaction.amount <= 0 {
            throw ValidationError.invalidAmount
        }
        
        if transaction.note.count > 250 {
            throw ValidationError.noteTooLong
        }
        
        if transaction.date > Date() {
            throw ValidationError.futureDate
        }
    }
}

File: android/app/src/main/kotlin/com/tripay/domain/DomainValidator.kt
package com.tripay.domain

import com.tripay.models.Transaction
import java.util.Date

sealed class ValidationError : Exception() {
    object InvalidAmount : ValidationError()
    object NoteTooLong : ValidationError()
    object FutureDate : ValidationError()
}

object DomainValidator {
    fun validate(transaction: Transaction) {
        if (transaction.amount <= 0) {
            throw ValidationError.InvalidAmount
        }

        if (transaction.note.length > 250) {
            throw ValidationError.NoteTooLong
        }

        if (transaction.date.after(Date())) {
            throw ValidationError.FutureDate
        }
    }
}
